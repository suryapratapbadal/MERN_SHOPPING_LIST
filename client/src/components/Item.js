import React, { Component } from 'react';

import IconButton from '@material-ui/core/IconButton';

import { withStyles } from '@material-ui/core/styles';
import ListItem from '@material-ui/core/ListItem';
import ListItemText from '@material-ui/core/ListItemText';
import EditIcon from '@material-ui/icons/Edit';
import SaveIcon from '@material-ui/icons/Save';
import Grid from '@material-ui/core/Grid';
import TextField from '@material-ui/core/TextField';
import ListItemSecondaryAction from '@material-ui/core/ListItemSecondaryAction';
import Checkbox from '@material-ui/core/Checkbox';

import Recipe from './Recipe';
import AddRecipe from './AddRecipe';
import DeleteConfermation from './DeleteConfermation';
import ToggleMenu from './ToggleMenu';

import { bindActionCreators } from 'redux';
import { connect } from 'react-redux';
import { updateItem } from '../actions/index';

class Item extends Component {

    constructor(props) {
        super(props);
        this.state = {
            editItem: false,
            itemName: '',
            toggleDeleteConfirmation: false,
            selected: false
        };
    }

    toggleDeleteConfirmationDialogue = () => {
        this.setState({ toggleDeleteConfirmation: !this.state.toggleDeleteConfirmation })
    }

    onChange = event => {
        this.setState({ itemName: event.target.value });
    }

    editItem = event => {
        event.preventDefault();
        event.stopPropagation();
        this.setState({ editItem: true, itemName: '' });
    }
    saveItem = event => {
        event.preventDefault();
        event.stopPropagation();
        if (this.state.itemName !== '') {
            this.props.updateItem(this.props._id, { "name": this.state.itemName });
            this.setState({ editItem: false });
        }
        else
            this.setState({ editItem: false });
    }


    render() {

        const { _id, name, completed, recipes, classes, loading } = this.props;

        return (
            <Grid container key={_id} className={loading ? classes.listItemDeactivate : classes.listItem}>
                <Grid item xs={12}>
                    <ListItem>
                        <Grid item xs={1} >
                            <Checkbox
                                onChange={() => this.setState({ selected: !this.state.selected })}
                                checked={this.state.selected}
                            />
                        </Grid>
                        <Grid item xs={6} className={completed ? classes.listItemDeactivate : ''}>
                            {
                                this.state.editItem ?
                                    <TextField
                                        id="standard-uncontrolled"
                                        label="Item Name"
                                        defaultValue={name}
                                        className={classes.textField}
                                        margin="normal"
                                        onChange={this.onChange}
                                        autoFocus
                                    /> :
                                    <ListItemText inset primary={name} />
                            }
                        </Grid>
                        <Grid item xs={1} className={completed ? classes.listItemDeactivate : ''}>
                            <IconButton
                                onClick={this.state.editItem ?  this.saveItem : this.editItem}
                                style={{ marginRight: '1rem' }}
                                disableRipple>
                                {
                                    this.state.editItem ?
                                        <SaveIcon /> :
                                        <EditIcon />
                                }

                            </IconButton>
                        </Grid>
                        <Grid item xs={4} className={completed ? classes.listItemDeactivate : ''}>
                            <AddRecipe id={_id} style={{ float: 'right' }} />
                        </Grid>

                        <ListItemSecondaryAction>
                            <DeleteConfermation toggleDeleteConfirmation={this.state.toggleDeleteConfirmation} id={_id} />
                            <ToggleMenu toggleDeleteConfirmationDialogue={this.toggleDeleteConfirmationDialogue} completed={completed} id={_id} />
                        </ListItemSecondaryAction>
                    </ListItem>
                </Grid>
                <Grid item xs={12} className={completed ? classes.listItemDeactivate : ''}>
                    <Recipe id={_id} recipes={recipes} />
                </Grid>
            </Grid>
        );
    }
}

const styles = theme => ({
    
    listItem: {
        marginBottom: '2rem',
        backgroundColor: theme.palette.background.paper,
    },
    listItemDeactivate: {
        opacity: 0.5,
        pointerEvents: 'none'
    },
    button: {
        margin: theme.spacing.unit,
    },
    textField: {
        marginLeft: theme.spacing.unit,
        marginRight: theme.spacing.unit,
        width: 150,
    }
});

export default connect(state => {
    const loading = state.itemReducer.loading || false;
    return {
        loading
    }
}, dispatch => {
    return bindActionCreators({ updateItem: updateItem }, dispatch)
}
)(withStyles(styles)(Item));