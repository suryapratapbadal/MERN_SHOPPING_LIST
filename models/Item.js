const mongoose = require('mongoose');
const Schema = mongoose.Schema;

// Create Schema
const ItemSchema = new Schema({
    name: {
        type: String,
        required: true
    },
    date: {
        type: Date,
        default: Date.now
    },
    completed: {
        type: Boolean,
        default: false
    }
});

const recipeTagsSchema = new Schema({
    item_id: {
        type: String,
        required: true
    },
    name: {
        type: String,
        required: true
    },
    date: {
        type: Date,
        default: Date.now
    },
});

Item = mongoose.model('item', ItemSchema);
Recipe = mongoose.model('recipe', recipeTagsSchema);
module.exports = {
    Item: Item,
    Recipe: Recipe
};