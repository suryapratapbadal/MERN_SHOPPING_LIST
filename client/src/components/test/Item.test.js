import React from 'react';
import { shallow } from 'enzyme';
import toJson from 'enzyme-to-json';

// Component to be tested
import { Item } from '../Item';


const mockupdatefn = jest.fn();
const recipes = [{ _id: '1', name: 'Tea' }];
const classes = {
    listItem: {
       
    },
    listItemDeactivate: {
        
    },
    button: {
        
    },
    textField: {
        
    }
};
const id ='1';


// Component Renders
describe('<Item />', () => {

    const wrapper = shallow(<Item recipes={recipes} classes={classes} updateItem={mockupdatefn} loading={false} _id={id} completed={false}/>);
    const event = {
        stopPropagation: jest.fn(),
        preventDefault: jest.fn(),
        target: { value : 'dummy'}
    }
    describe('render()', () => {

        test('renders the component', () => {

            expect(toJson(wrapper)).toMatchSnapshot();

        });
    });


    describe('onChange()',()=>{
        test('handle click on button',()=>{
            wrapper.find('WithStyles(Checkbox)').props().onChange();
            expect(wrapper.state('selected')).toEqual(true);
        });
    });

    // describe('onChange()',()=>{
    //     test('handle click on button',()=>{
    //         wrapper.find('WithStyles(TextField)').at(0).props().onChange(event);
    //         expect(wrapper.state('itemName')).toEqual('dummy');
    //     });
    // });

    // describe('onClose()',()=>{
    //     test('handle click on button',()=>{
    //         wrapper.find('WithStyles(Menu)').props().onClose();
    //         expect(wrapper.state('anchorEl')).toEqual(null);
    //     });
    // });

    describe('onClick()',()=>{
        test('handle click on button',()=>{
            wrapper.find('WithStyles(IconButton)').at(0).props().onClick(event);
            expect(wrapper.state('editItem')).toEqual(true);
        });
    });
});