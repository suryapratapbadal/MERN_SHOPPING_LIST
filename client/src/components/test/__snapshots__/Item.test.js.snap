// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Item /> render() renders the component 1`] = `
<WithStyles(Grid)
  className={Object {}}
  container={true}
  key="1"
>
  <WithStyles(Grid)
    item={true}
    xs={12}
  >
    <WithStyles(ListItem)>
      <WithStyles(Grid)
        item={true}
        xs={1}
      >
        <WithStyles(Checkbox)
          checked={false}
          onChange={[Function]}
        />
      </WithStyles(Grid)>
      <WithStyles(Grid)
        className=""
        item={true}
        xs={6}
      >
        <WithStyles(ListItemText)
          inset={true}
        />
      </WithStyles(Grid)>
      <WithStyles(Grid)
        className=""
        item={true}
        xs={1}
      >
        <WithStyles(IconButton)
          disableRipple={true}
          onClick={[Function]}
          style={
            Object {
              "marginRight": "1rem",
            }
          }
        >
          <pure(EditIcon) />
        </WithStyles(IconButton)>
      </WithStyles(Grid)>
      <WithStyles(Grid)
        className=""
        item={true}
        xs={4}
      >
        <Connect(WithStyles(AddRecipe))
          id="1"
          style={
            Object {
              "float": "right",
            }
          }
        />
      </WithStyles(Grid)>
      <WithStyles(ListItemSecondaryAction)>
        <Connect(DeleteConfermation)
          id="1"
          toggleDeleteConfirmation={false}
        />
        <Connect(ToggleMenu)
          completed={false}
          id="1"
          toggleDeleteConfirmationDialogue={[Function]}
        />
      </WithStyles(ListItemSecondaryAction)>
    </WithStyles(ListItem)>
  </WithStyles(Grid)>
  <WithStyles(Grid)
    className=""
    item={true}
    xs={12}
  >
    <Connect(WithStyles(Recipe))
      id="1"
      recipes={
        Array [
          Object {
            "_id": "1",
            "name": "Tea",
          },
        ]
      }
    />
  </WithStyles(Grid)>
</WithStyles(Grid)>
`;
